presto-main/src/main/java/io/prestosql/sql/planner/DomainCoercer.java

```diff
@@ -0,0 +1,242 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner;
+
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableList;
+import io.prestosql.Session;
+import io.prestosql.metadata.Metadata;
+import io.prestosql.metadata.OperatorNotFoundException;
+import io.prestosql.metadata.ResolvedFunction;
+import io.prestosql.spi.PrestoException;
+import io.prestosql.spi.connector.ConnectorSession;
+import io.prestosql.spi.function.InvocationConvention;
+import io.prestosql.spi.predicate.AllOrNoneValueSet;
+import io.prestosql.spi.predicate.Domain;
+import io.prestosql.spi.predicate.Marker;
+import io.prestosql.spi.predicate.Range;
+import io.prestosql.spi.predicate.Ranges;
+import io.prestosql.spi.predicate.SortedRangeSet;
+import io.prestosql.spi.predicate.ValueSet;
+import io.prestosql.spi.type.Type;
+import io.prestosql.spi.type.TypeOperators;
+import io.prestosql.sql.InterpretedFunctionInvoker;
+
+import java.lang.invoke.MethodHandle;
+import java.util.Optional;
+
+import static io.prestosql.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static io.prestosql.spi.function.InvocationConvention.InvocationArgumentConvention.NEVER_NULL;
+import static io.prestosql.spi.function.InvocationConvention.InvocationReturnConvention.FAIL_ON_NULL;
+import static io.prestosql.spi.function.OperatorType.SATURATED_FLOOR_CAST;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Apply saturated floor casts for implicit coercions on TupleDomain
+ */
+public final class DomainCoercer
```

Comment:
add a not that this doesn't handle `NaNs` since it's being used by dynamic filtering and `NaNs` are not possible

Reply:
We should also have explicit check for NaN